docs/designs/odata-writer-rewrite-proposal.md

```diff
@@ -0,0 +1,315 @@
+
+- You can configure various settings and options in `ODataMessageWriterSettings`
+- You can replace the low-level `IJsonWriter` by implementing the interface
+- You can control what gets written through the properties and annotatons that you add to `ODataResource` and similar types
+
+Here's a rough proposal for the new writer:
+
+```csharp
+interface IODataMessageWriter<TContext, TState>
+{
+    IODataWriter<TContext, TState, TValue> GetWriter<TValue>(ODataPayloadKind payloadKind);
+    // Alternatively we could have different `GetWriter` methods for different
+}
+
+
+// Writer for a specific payload kind
+interface IODataWriter<TContext, TState, TValue>
+{
+    ValueTask WriterAsync(TValue value, TState state, TContext context);
+}
+```
+
+where:
+
+- `TContext`: represents an implementation-specific context type that encapsulates session-specific settings and format information (e.g. JSON options, ODataUri, Output stream, custom services)
+- `TState`: implementation-specific state type that represents the current state of the serialization. Unlike the context, this is expected to change per object or level (e.g. current type being serialized, nesting depth, etc.). I'm not yet sure if it makes sense to have this
+- `TValue`: the value being written. This is generic so we can write any value and make it possible to create strongly-typed writers without boxing, or without heavy use of reflection (e.g. source generators)
+
+This is a fairly high-level generic that I'm still debating whether it's too generic to be useful. We do not expect customers to create custom implementation for this interface,
+but it gives us some leeway to implement a writer for a different a output format if necessary. In practice, we would create a factory that returns a concrete implementation
+based on some user configuration or builder.
+
+We would have a single default `IODataMessageWriter` implementation for the default formats we support (xml $metadata, JSON paylods, raw values), and some default `IODataWriter` implementations
+for different payload types (for resources, resource sets, etc.). The default `IODataWriter` implementation would support extensibility via custom provider interfaces to provide support
+for different data types (e.g. CLR POCO classes, custom input sources, etc.)
+
+Alternatively, we could have a single writer interface and create factory class that can create a factory or builder that can return a concrete instance based on configuration:
+
+```csharp
+interface IODataWriter<TContext, TState, TValue>
+{
+    ValueTask WriteAsync(TValue value, TState state, TContext context);
+}
+
+var odataWriter = ODataWriterBuilder.FromPayloadKind(payloadKind)
+    .WithJsonOptions(jsonSerializationOptions)
+    .WithODataUri(odataUri)
+    .ConfigureResourceWriters(provider => provider.Add<Customer, CustomerResourceWriter>())
+    .ForCollectionOf<Customer>()

```

Comment:
Let's remember that AGS wants to use our readers and writers, and doesn't have CLR types for the types represented in the model

Reply:
In the case of AGS they would inject custom writers for their input types `JArray`, `JObject` or the input stream. The build pattern here is to provide a discoverable way to configure the service, so there could be helper methods for different common use cases (including when it's not an `IEnumerable<T>`), while also providing where to define the type directly for the less common use cases.